src/compiler/path.ts

```diff
@@ -624,12 +624,115 @@ export function getNormalizedPathComponents(path: string, currentDirectory: stri
-export function getNormalizedAbsolutePath(fileName: string, currentDirectory: string | undefined): string {
-    return getPathFromPathComponents(getNormalizedPathComponents(fileName, currentDirectory));
+export function getNormalizedAbsolutePath(path: string, currentDirectory: string | undefined): string {
+    let rootLength = getRootLength(path);
+    if (rootLength === 0 && currentDirectory) {
+        path = combinePaths(currentDirectory, path);
+        rootLength = getRootLength(path);
+    }
+    const simple = simpleNormalizePath(path);
+    if (simple !== undefined) {
+        return simple;
+    }
+    const root = path.substring(0, rootLength);
+    const normalizedRoot = root && normalizeSlashes(root);
+    // `normalized` is only initialized once `path` is determined to be non-normalized
+    let normalized = normalizedRoot === root ? undefined : normalizedRoot;
+    let index = rootLength;
+    let segmentStart = index;
+    let normalizedUpTo = index;
+    let seenNonDotDotSegment = rootLength !== 0;
+    while (index < path.length) {
+        // At beginning of segment
+        segmentStart = index;
+        let ch = path.charCodeAt(index);
+        while (isAnyDirectorySeparator(ch) && index + 1 < path.length) {
+            index++;
+            ch = path.charCodeAt(index);
+        }
+        if (index > segmentStart) {
+            if (normalized === undefined) {
+                // Seen superfluous separator
+                normalized = path.substring(0, segmentStart - 1);
+            }
+            segmentStart = index;
+        }
+        // Past any superfluous separators
+        const sepIndex = path.indexOf(directorySeparator, index + 1);
+        const altSepIndex = path.indexOf(altDirectorySeparator, index + 1);
+        let segmentEnd = sepIndex === -1 ? altSepIndex : altSepIndex === -1 ? sepIndex : Math.min(sepIndex, altSepIndex);
+        if (segmentEnd === -1) {
+            segmentEnd = path.length;
+        }
+        if (segmentEnd === altSepIndex && normalized === undefined) {
+            // Seen backslash
+            normalized = path.substring(0, segmentStart);
+        }
+        const segmentLength = segmentEnd - segmentStart;
+        if (segmentLength === 1 && path.charCodeAt(index) === CharacterCodes.dot) {
+            // "." segment (skip)
+            if (normalized === undefined) {
+                normalized = path.substring(0, normalizedUpTo);
+            }
+        }
+        else if (segmentLength === 2 && path.charCodeAt(index) === CharacterCodes.dot && path.charCodeAt(index + 1) === CharacterCodes.dot) {
+            // ".." segment
+            if (!seenNonDotDotSegment) {
+                if (normalized !== undefined) {
+                    normalized += normalized.length === rootLength ? ".." : "/..";
+                }
+                else {
+                    normalizedUpTo = index + 2;
+                }
+            }
+            else if (normalized === undefined) {

```

Comment:
I think that in this branch it is *technically* possible for you to avoid extra substrings by adjusting `normalizedUpTo` instead of initializing `normalized`.